src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+extension method and its `on` type.
+Members that apply to the supertype are applicable to the subtype,
+unless the supertype has a more specific definition for the member.
+
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
+is not an override relationship for extension types, but a *redeclaration*.
+Use the `@redeclare` annotation to tell the compiler you are *knowingly* choosing
+to use the same name as a supertype's member definition, not just accidentally
+using the same name:
+
+```dart
+class C {
+  void m() {}
+}
+
+extension type E(C c) implements C {
+  @redeclare
+  void m() {}
+}
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
+
+## Usage
+
+To use an extension type, create an instance the same as you would with a class:
+
+```dart
+extension type NumberE(int value) {
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+
+  NumberE get i => value;
+  bool isValid() => !value.isNegative;
+}
+
+void testE() { 
+  var num = NumberE(1);
+}
+```
+
+Then you can invoke members on the object as you would with a class object.
+Any attempt to treat the extension type object as an object
+of the representation type, or vice versa, will result in compile time errors
+(unless that behavior is explicitly defined for the extension type,
+or it is [transparent](#transparency)).
+For example:
+
+```dart
+void testE() { 
+  var num = NumberE(1);
+  int num2 = NumberE(2); // Error: Can't assign 'NumberE' to 'int'.
+  
+  num.isValid(); // Ok: Extension member invocation.
+  num.isNegative(); // Error: 'NumberE' does not define 'int' member 'isNegative'.
+  
+  var sum = num + num; // Ok: 'NumberE' defines '+'.
+  var diff = num - num; // Error: 'NumberE' does not define 'int' member '-'.
+  var diff2 = num.value - 2; // Ok: Can access representation object with reference.
+  var sum2 = num + 2; // Error: Can't assign 'int' to parameter type 'NumberE'. 
+  
+  List<NumberE> numbers = [
+    NumberE(1), 
+    num.i, // Ok: 'i' getter returns type 'NumberE'.
+    1, // Error: Can't assign 'int' element to list type 'NumberE'.
+    1 as NumberE, // Ok: Dynamic type cast recognizes representation type.
+  ];
+}
+```
+
+#### Transparency
+
+An extension type that [implements](#implements) its representation type
+is considered *transparent* because it gives the extension type access
+to *all* members of the representation type. Otherwise,
+only the members declared in the extension type body are available to it
+(considered a [*somewhat protected*](#type-considerations) extension type).
+
+```dart
+extension type NumberT(int value) 
+  implements int {
+  // Doesn't explicitly declare any members of 'int'.
+  NumberT get i => NumberT(value);
+}
+
+void main () {
+  var num = NumberT(1);
+  
+  // All OK: Representation type interface is available to extension type:
```

Comment:
Perhaps?:
```suggestion
  // All OK: `implements int` allows using `int` members on the extension type:
```

The point is that the decision is made based on the type in the implements clause. If we had written `implements num` then we would not be able to invoke `isEven` on an expression of type `NumberT`, even in the case where the representation type is `int`.

There is an extra constraint (that you already mention elsewhere): If we have `implements T` where `T` is a non-extension type then it must be true that the representation type `R` is a subtype of `T`.

It's probably going to be rather common that an extension type uses the same non-extension type `T` as (1) the representation type and (2) as a superinterface (as in `implements ...T...`), but there are other possibilities.